source_code/main_mcu/src/LOGIC/logic_gui.c

```diff
@@ -303,23 +333,36 @@ void logic_gui_display_login_password(child_cred_node_t* child_node)
         utils_ascii_to_unicode(child_node->password, NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH);
         child_node->cust_char_password[NODEMGMT_OLD_GEN_ASCII_PWD_LENGTH] = 0;
     }    
-    
+
+    /* Decrypt TOTPsecret */
+    logic_encryption_ctr_decrypt(child_node->TOTP.TOTPsecret_ct, child_node->TOTP.TOTPsecret_ctr, MEMBER_SIZE(TOTP_cred_node_t, TOTPsecret), FALSE);
+
     /* Temp vars for our main loop */
     BOOL first_function_run = TRUE;
-    int16_t text_anim_x_offset[2];
-    BOOL text_anim_going_right[2];
+
+    int16_t text_anim_x_offset[LOGIC_GUI_NUM_LINES];
+    BOOL text_anim_going_right[LOGIC_GUI_NUM_LINES];
     BOOL redraw_needed = TRUE;
     int16_t displayed_length;
-    BOOL scrolling_needed[2];
+    BOOL scrolling_needed[LOGIC_GUI_NUM_LINES];
     
     /* Reset temp vars */
     memset(text_anim_going_right, FALSE, sizeof(text_anim_going_right));
     memset(text_anim_x_offset, 0, sizeof(text_anim_x_offset));
     memset(scrolling_needed, FALSE, sizeof(scrolling_needed));
     
+    /* Generate TOTP */
+    cust_char_t TOTP_str[LOGIC_GUI_TOTP_STR_LEN];
+    memset(TOTP_str, 0, sizeof TOTP_str);
+    if (child_node->TOTP.TOTPsecretLen > 0)
+    {
+        uint8_t remaining_secs = logic_encryption_generate_totp(child_node->TOTP.TOTPsecret_ct, child_node->TOTP.TOTPsecretLen, child_node->TOTP.TOTPnumDigits, TOTP_str, LOGIC_GUI_TOTP_STR_LEN);
+        logic_gui_create_TOTP_display_str(TOTP_str, LOGIC_GUI_TOTP_STR_LEN, child_node->TOTP.TOTPnumDigits, remaining_secs);
+    }
+
     /* Lines display settings */
-    cust_char_t* strings_to_be_displayed[2] = {child_node->login, child_node->cust_char_password};
-    uint16_t strings_y_positions[4] = {12, 36};
+    cust_char_t* strings_to_be_displayed[LOGIC_GUI_NUM_LINES] = {child_node->login, child_node->cust_char_password, TOTP_str};
+    uint16_t strings_y_positions[4] = {5, 20, 35};
```

Comment:
these lines need to be changed to handle the 2 scenarios: with and without TOTP :)

Reply:
ok, will change. I was just going to leave it showing a gaping hole :-). Display would shows this currently:

With TOTP:
<login>
<pwd>
<TOTP>

Without TOTP:
<login>
<pwd>

With TOTP but no password:
<login>

<TOTP>

Reply:
any way to do the same presentation for "without totp" and "with totp but no password" ? :)